src/normalize.rs

```diff
@@ -1,22 +1,51 @@
 use crate::{Error, RResult, ResolveResult, Resolver};
 
 use path_absolutize::Absolutize;
-use std::path::{Path, PathBuf};
 
-impl Resolver {
-    fn normalize_path_without_link(path: &Path) -> PathBuf {
+use std::{
+    borrow::Cow,
+    path::{Path, PathBuf},
+};
+
+pub trait NormalizePath {
+    fn normalize(&self) -> Cow<Path>;
+}
+
+impl NormalizePath for Path {
+    #[inline]
+    fn normalize(&self) -> Cow<Path> {
         // perf: this method does not re-allocate memory if the path does not contain any dots.
-        path.absolutize_from(Path::new("")).unwrap().to_path_buf()
+        self.absolutize_from(Path::new("")).unwrap()
     }
+}
 
+impl Resolver {
     #[tracing::instrument]
     fn normalize_path(&self, path: &Path) -> RResult<PathBuf> {
-        let path = Self::normalize_path_without_link(path);
         if self.options.symlinks {
-            let entry = self.load_entry(&path)?;
-            entry.symlink().map_err(Error::Io)
+            let mut entry = self.load_entry(path)?;
+            let has_symlink = loop {
+                if entry
+                    .stat
+                    .read()
+                    .unwrap()
+                    .map_or(false, |stat| stat.is_symlink)
+                {
+                    break true;
+                }
+                if let Some(e) = entry.parent.clone() {
+                    entry = e;
+                } else {
+                    break false;
+                }
+            };
+            if has_symlink {
+                entry.symlink().map_err(Error::Io)
+            } else {
+                Ok(path.normalize().to_path_buf())
+            }
```

Comment:
@bvanjoi is this legit? Test if any of the parent paths contain a symlink, and only call `canonicalize` if there exists a symlink.

Reply:
I'll checkout the tests tomorrow

Reply:
maybe not, for example: `/linked/this/outer/` pointed to `/linked`, we need to the real path when resolve `linked/this/outer/lib/index.js`.

<img width="347" alt="image" src="https://user-images.githubusercontent.com/30187863/215328638-660bfa5a-7635-435f-9c4f-3a4ee6114b4e.png">